File: content/tracks/dsa-javascript/concepts/01-two-pointers/two-pointers_test.js
const assert = require("assert");
const { isPalindrome, containerWithMostWater, threeSum } = require("./stub");

let passed = 0, failed = 0;

function test(name, fn) {
  try {
    fn();
    console.log(`  PASS: ${name}`);
    passed++;
  } catch (e) {
    console.log(`  FAIL: ${name} — ${e.message}`);
    failed++;
  }
}

// Helper: sort triplets for comparison
function sortTriplets(arr) {
  return arr.map(t => [...t].sort((a, b) => a - b)).sort((a, b) => {
    for (let i = 0; i < 3; i++) if (a[i] !== b[i]) return a[i] - b[i];
    return 0;
  });
}

// isPalindrome
test("isPalindrome: classic phrase", () => {
  assert.strictEqual(isPalindrome("A man, a plan, a canal: Panama"), true);
});
test("isPalindrome: not a palindrome", () => {
  assert.strictEqual(isPalindrome("race a car"), false);
});
test("isPalindrome: single space", () => {
  assert.strictEqual(isPalindrome(" "), true);
});
test("isPalindrome: empty string", () => {
  assert.strictEqual(isPalindrome(""), true);
});
test("isPalindrome: simple word", () => {
  assert.strictEqual(isPalindrome("racecar"), true);
});
test("isPalindrome: numbers only", () => {
  assert.strictEqual(isPalindrome("12321"), true);
});
test("isPalindrome: mixed punctuation palindrome", () => {
  assert.strictEqual(isPalindrome("Was it a car or a cat I saw?"), true);
});
test("isPalindrome: single character", () => {
  assert.strictEqual(isPalindrome("a"), true);
});
test("isPalindrome: two different chars", () => {
  assert.strictEqual(isPalindrome("ab"), false);
});

// containerWithMostWater
test("containerWithMostWater: classic example", () => {
  assert.strictEqual(containerWithMostWater([1, 8, 6, 2, 5, 4, 8, 3, 7]), 49);
});
test("containerWithMostWater: two elements equal", () => {
  assert.strictEqual(containerWithMostWater([1, 1]), 1);
});
test("containerWithMostWater: two elements unequal", () => {
  assert.strictEqual(containerWithMostWater([4, 3]), 3);
});
test("containerWithMostWater: ascending heights", () => {
  assert.strictEqual(containerWithMostWater([1, 2, 3, 4, 5]), 6);
});
test("containerWithMostWater: all same height", () => {
  assert.strictEqual(containerWithMostWater([5, 5, 5, 5]), 15);
});
test("containerWithMostWater: best pair not at ends", () => {
  assert.strictEqual(containerWithMostWater([1, 5, 4, 5, 1]), 12);
});
test("containerWithMostWater: includes zero height", () => {
  assert.strictEqual(containerWithMostWater([0, 2, 0]), 0);
});

// threeSum
test("threeSum: classic example", () => {
  const result = sortTriplets(threeSum([-1, 0, 1, 2, -1, -4]));
  const expected = sortTriplets([[-1, -1, 2], [-1, 0, 1]]);
  assert.deepStrictEqual(result, expected);
});
test("threeSum: no valid triplets", () => {
  assert.deepStrictEqual(threeSum([0, 1, 1]), []);
});
test("threeSum: all zeros", () => {
  assert.deepStrictEqual(sortTriplets(threeSum([0, 0, 0])), [[0, 0, 0]]);
});
test("threeSum: multiple duplicates", () => {
  const result = sortTriplets(threeSum([-2, 0, 0, 2, 2]));
  assert.deepStrictEqual(result, [[-2, 0, 2]]);
});
test("threeSum: empty array", () => {
  assert.deepStrictEqual(threeSum([]), []);
});
test("threeSum: two elements", () => {
  assert.deepStrictEqual(threeSum([0, 0]), []);
});
test("threeSum: all positive no solution", () => {
  assert.deepStrictEqual(threeSum([1, 2, 3, 4]), []);
});
test("threeSum: all negative no solution", () => {
  assert.deepStrictEqual(threeSum([-4, -3, -2, -1]), []);
});

console.log(`\n${passed} passed, ${failed} failed`);
if (failed > 0) process.exit(1);

File: content/tracks/dsa-python/concepts/01-two-pointers/two-pointers_test.py
import sys
from stub import is_palindrome, container_with_most_water, three_sum

passed = 0
failed = 0


def test(name: str, condition: bool, detail: str = "") -> None:
    global passed, failed
    if condition:
        print(f"  PASS: {name}")
        passed += 1
    else:
        msg = f" — {detail}" if detail else ""
        print(f"  FAIL: {name}{msg}")
        failed += 1


# is_palindrome
test("palindrome: classic phrase", is_palindrome("A man, a plan, a canal: Panama") is True)
test("palindrome: not palindrome", is_palindrome("race a car") is False)
test("palindrome: empty string", is_palindrome("") is True)
test("palindrome: single char", is_palindrome("a") is True)
test("palindrome: digits", is_palindrome("12321") is True)
test("palindrome: digits false", is_palindrome("12345") is False)
test("palindrome: punctuation only", is_palindrome(".,!") is True)
test("palindrome: no lemon no melon", is_palindrome("No lemon, no melon") is True)

# container_with_most_water
r = container_with_most_water([1, 8, 6, 2, 5, 4, 8, 3, 7])
test("container: standard", r == 49, f"got {r}")
r = container_with_most_water([1, 1])
test("container: two equal", r == 1, f"got {r}")
r = container_with_most_water([4, 3, 2, 1, 4])
test("container: equal ends", r == 16, f"got {r}")
r = container_with_most_water([1, 2, 1])
test("container: three elements", r == 2, f"got {r}")
r = container_with_most_water([2, 3, 4, 5, 18, 17, 6])
test("container: best not at ends", r == 17 * 4, f"got {r}")

# three_sum
def sorted_triplets(lst: list[list[int]]) -> list[list[int]]:
    return sorted(sorted(t) for t in lst)

r = sorted_triplets(three_sum([-1, 0, 1, 2, -1, -4]))
test("three_sum: standard", r == [[-1, -1, 2], [-1, 0, 1]], f"got {r}")

r = sorted_triplets(three_sum([0, 0, 0]))
test("three_sum: all zeros", r == [[0, 0, 0]], f"got {r}")

r = sorted_triplets(three_sum([0, 1, 1]))
test("three_sum: no solution", r == [], f"got {r}")

r = sorted_triplets(three_sum([-2, 0, 0, 2, 2]))
test("three_sum: duplicates in input", r == [[-2, 0, 2]], f"got {r}")

r = sorted_triplets(three_sum([1, 2, 3]))
test("three_sum: no negative", r == [], f"got {r}")

r = sorted_triplets(three_sum([0, 0, 0, 0]))
test("three_sum: four zeros", r == [[0, 0, 0]], f"got {r}")

print(f"\n{passed} passed, {failed} failed")
if failed > 0:
    sys.exit(1)
